### ISSUE TITLE
errors on installing on Vim 8.0.1831 and below

### ISSUE THREAD

**Issue by dradtke** [fatih/vim-go#2520]
### What did you do? (required: The issue will be **closed** when not provided)

While installing `vim-go` in a fresh development environment, I ran into a couple issues. I installed the plugin using vim-plug like this:

```
Plug 'fatih/vim-go', { 'do': ':GoInstallBinaries' }
```

and when I run `:PlugInstall` the first time, I get this:

```
x Post-update hook for vim-go ... Vim(unlet):E488: Trailing characters
```

Re-running installation indicates that the plugin is already installed, so it seems to have worked despite the issue.

Then when I open a Go file, I get this error:

```
vim-go: could not find 'gopls'. Run :GoInstallBinaries to fix it                                                                                                                                                                              
Error detected while processing function <SNR>19_register[9]..go#lsp#DidOpen:                                                                                                                                                                 
line   19:                                                                                                                                                                                                                                    
E117: Unknown function: sendMessage                                                                                                                                                                                                           
E15: Invalid expression: l:lsp.sendMessage(l:msg, l:state)
```

Odd, because it should have already attempted to install binaries. Re-running `:GoInstallBinaries` manually does seem to fix it, but not after outputting another error, whose text I can't grab because it quickly gets overwritten by further installation messages.

### What did you expect to happen?

Everything to install successfully with no errors.

### What happened instead?

Everything looks like it installed successfully, but only after several error messages, and having to install binaries several times.

### Configuration (**MUST** fill this out):

#### vim-go version: latest (commit `a7b5edc1872b76e6c0cc18b89fcda3aa98422961`)

####  `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):
<details><summary>vimrc</summary><br><pre>
call plug#begin()
Plug 'fatih/vim-go', { 'do': ':GoInstallBinaries' }
call plug#end()
</pre></details>


#### Vim version (first three lines from `:version`):
<!-- :version -->

```
VIM - Vi IMproved 8.0 (2016 Sep 12)                                                                                    
Included patches: 1-1568                                                                                               
Compiled by 'http://www.opensuse.org/'
```

#### Go version (`go version`):
<!-- go version -->

```
go version go1.13.1 linux/amd64
```

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>
<!-- go env -->
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/damien/.cache/go-build"
GOENV="/home/damien/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/damien/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build736621601=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

**Comment by bhcleek** [comment]
Given that `:GoInstallBinaries` works, but the post hook doesn't, it seems like this _might_ be an issue in Plug, but let's rule issues with vim-go to be sure. Can you reliably duplicate the problem?

**Comment by dradtke** [comment]
Yep, and it looks like it's not an issue with Plug. If I run `rm -rf ~/go/bin` and then re-open a Go file in vim, I get

```
vim-go: could not find 'gopls'. Run :GoInstallBinaries to fix it                                                       
Error detected while processing function <SNR>19_register[9]..go#lsp#DidOpen:                                          
line   19:                                                                                                             
E117: Unknown function: sendMessage                                                                                    
E15: Invalid expression: l:lsp.sendMessage(l:msg, l:state)
```

The first line makes sense, but the rest of it seems odd. Running `:GoInstallBinaries` then works, but it first prints out a "trailing characters" error.

**Comment by bhcleek** [comment]
yeah, the first line makes sense, and I can take care of the rest of it once we know what's actually wrong here.

Can you see if #2521 solves the issue for you? If it does, then you should also have some output that will tell us more about what's wrong. If it goes by faster than you can read it, `:messages` will show the messages I need.

**Comment by dradtke** [comment]
Looks like it didn't solve it, but here are the messages:

```
vim-go: could not find 'gopls'. Run :GoInstallBinaries to fix it                                                       
Error detected while processing function <SNR>19_register[9]..go#lsp#DidOpen:                                          
line   19:                                                                                                             
E117: Unknown function: sendMessage                                                                                    
E15: Invalid expression: l:lsp.sendMessage(l:msg, l:state)                                                             
vim-go: could not unset $GO111MODULE: Vim(unlet):E488: Trailing characters
```

**Comment by dradtke** [comment]
I just confirmed that this seems to be the line it's complaining about, at least as far as the `go#lsp#DidOpen` error: https://github.com/fatih/vim-go/blob/a7b5edc1872b76e6c0cc18b89fcda3aa98422961/autoload/go/lsp.vim#L508

**Comment by dradtke** [comment]
Even with that commented out, I still get the `could not unset` error though, so the two don't seem to be directly related.

**Comment by bhcleek** [comment]
The issue with `sendMessage` is a red herring. I'm aware of where that line is, but the error is just a consequence of something else.

The messages you most recently reported don't seem to actually match what you originally reported. Can you duplicate the `x Post-update hook for vim-go ... Vim(unlet):E488: Trailing characters` message?

**Comment by dradtke** [comment]
It looks like that `sendMessage` error is likely to happen anywhere `gopls` isn't already installed, since it returns a blank `l:lsp` object here, but callers of this function seem to treat it as a successful result. https://github.com/fatih/vim-go/blob/a7b5edc1872b76e6c0cc18b89fcda3aa98422961/autoload/go/lsp.vim#L363

Yeah, the post-update hook error is easily repeatable by deleting both `~/go/bin` and `~/.vim/plugged` and then re-running `PlugInstall`.

If I remove `, { 'do': ':GoInstallBinaries' }` from the vimrc, then everything installs fine, but I get the same errors from vim-go when I then quit, open a Go file, and run `:GoInstallBinaries` manually:

```
vim-go: could not find 'gopls'. Run :GoInstallBinaries to fix it                                                                                                                                                                              
Error detected while processing function <SNR>19_register[9]..go#lsp#DidOpen:                                                                                                                                                                 
line   19:                                                                                                                                                                                                                                    
E117: Unknown function: sendMessage                                                                                                                                                                                                           
E15: Invalid expression: l:lsp.sendMessage(l:msg, l:state)                                                                                                                                                                                    
Error detected while processing function <SNR>19_GoInstallBinaries[126]..<SNR>21_remove:                                                                                                                                                      
line    1:                                                                                                                                                                                                                                    
E488: Trailing characters
````

**Comment by bhcleek** [comment]
Do you have the problem when you:
* exit vim
* delete ~/go/bin and ~/.vim/pluggged
* open vim (don't try to open a Go file!)
* execute `:GoInstallBinaries`
* open a Go file.

**Comment by bhcleek** [comment]
I've pushed fixes to #2521 that will allow us to focus on the problems you're having with vim-plug.

**Comment by dradtke** [comment]
Ah, it looks like `unlet $VAR` is in a newer version of Vim than what I'm running. I have patches 1-1568, and it looks like the ability to unlet environment variables was introduced in 1832. https://github.com/vim/vim/issues/1116

Thanks for the update, I'll give it another shot with your change to see if that helps the lsp issue.

**Comment by dradtke** [comment]
Yep, with those changes the only errors I get are `could not find 'gopls'` (which is expected) and `could not unset $GO111MODULE`, which is unfortunate but also kind of expected.

Interestingly, now it tries to initialize gopls twice:

```
vim-go: could not find 'gopls'. Run :GoInstallBinaries to fix it                                                       
vim-go: initializing gopls                                                                                             
vim-go: could not unset $GO111MODULE: Vim(unlet):E488: Trailing characters                                             
vim-go: initializing gopls                                                                                             
vim-go: installing finished!                                                                                           
vim-go: initialized gopls                                                                                              
vim-go: [exit] SUCCESS                                                                                                 
vim-go: initializing gopls                                                                                             
vim-go: initialized gopls
```

This is still a lot better than it was, though.

**Comment by bhcleek** [comment]
Thanks for the link to the Vim issue; that makes sense.

Yes, it's going to keep trying to initialize gopls until it's available.

**Comment by bhcleek** [comment]
Fixed in #2523 and #2522 .

Thanks for the report and for helping me figure out the problem with Vim 8.0.1831 and below.
